File: Data Structures & Algorithms/level-order-traversal-of-binary-tree/submission-4.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    vector<vector<int>> levelOrder(TreeNode* root) {
        vector<vector<int>> res; 
        if (root == nullptr) return res;

        std::queue<TreeNode*> q;
        q.push(root);

        while(!q.empty()){
            vector<int> curLevel;
            int size = q.size();
            
            for(int i = 0; i < size; i++){
                TreeNode* node = q.front(); q.pop();
                if(node){
                    curLevel.push_back(node->val);
                    q.push(node->left);
                    q.push(node->right);
                }
            }
            if(!curLevel.empty()){
                res.push_back(curLevel);
            }
        }
        return res;
    }
};

File: Data Structures & Algorithms/populating-next-right-pointers-in-each-node/submission-4.js
/**
 * Definition for a binary tree node.
 * class Node {
 *     constructor(val = 0, left = null, right = null, next = null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *         this.next = next;
 *     }
 * }
 */

class Solution {
    /**
     * @param {Node} root
     * @return {Node}
     */
    connect(root) {
        if (!root) return null;
        let res = root;
        let q = [root];
        while (q.length > 0) {
            let len = q.length;
            for (let i = 0; i < len; i++) {
                let val = q.shift();
                if (i < len - 1) {
                    val.next = q[0];
                }
                if (val.left) q.push(val.left);
                if (val.right) q.push(val.right);
            }
        }
        return res;
    }
}
